clang/lib/Sema/SemaExpr.cpp

```diff
@@ -5915,6 +5916,16 @@ bool Sema::GatherArgumentsForCall(SourceLocation CallLoc, FunctionDecl *FDecl,
           ProtoArgType->isBlockPointerType())
         if (auto *BE = dyn_cast<BlockExpr>(Arg->IgnoreParenNoopCasts(Context)))
           BE->getBlockDecl()->setDoesNotEscape();
+      // During template instantiation it is possible that we already created
+      // the HLSLOutArgExpr if it was produced during tree transformation.
```

Comment:
Generally, `TreeTransform` just drops these implicit nodes on the floor, instantiates the original expression, and creates new nodes when it's re-processing the instantiated call.  This can get complicated around OVEs for e.g. pseudo-object expressions, but in your case, it should be easy because *all* of the extra structure is implicit stuff, and we haven't restructured anything.

Reply:
Since you've fixed TreeTransform to look through the `HLSLOutArgExpr`, it should now no longer be possible to see an `HLSLOutArgExpr` here.